staging/src/k8s.io/client-go/tools/cache/thread_safe_store.go

```diff
@@ -148,12 +148,20 @@ func (c *threadSafeMap) Index(indexName string, obj interface{}) ([]interface{},
 	}
 	index := c.indices[indexName]
 
-	// need to de-dupe the return list.  Since multiple keys are allowed, this can happen.
-	returnKeySet := sets.String{}
-	for _, indexKey := range indexKeys {
-		set := index[indexKey]
-		for _, key := range set.UnsortedList() {
-			returnKeySet.Insert(key)
+	var returnKeySet sets.String
+	if len(indexKeys) == 1 {
+		// In majority of cases, there is exactly one value matching.
+		// Optimize the most common path - deduping is not needed here.
+		returnKeySet = index[indexKeys[0]]
+	} else {
+		// Need to de-dupe the return list.
+		// Since multiple keys are allowed, this can happen.
+		returnKeySet = sets.String{}
+		for _, indexKey := range indexKeys {
+			set := index[indexKey]
+			for _, key := range set.UnsortedList() {
```

Comment:
can also iterate over the set directly (`for k := range set`), rather than forcing an allocation by calling UnsortedList

Reply:
We don't return it - it's used couple lines ago (still under lock) to restore values of the items.

Reply:
done